File: loglib-java/src/main/java/com/robog/loglib/DatabaseLog.java
package com.robog.loglib;


import android.content.Context;
import android.os.Process;
import android.util.Log;

import com.robog.loglib.database.LogBeanDao;
import com.robog.loglib.database.LogBeanDaoImpl;

import java.util.List;

import static com.robog.loglib.SLog.TAG;

/**
 * Created by yuxingdong on 2018/6/21.
 * <p>
 * 将Log对象存入数据库，并开启数据库监测，定期检查数据库数据数量是否达到上限。
 * </p>
 */
final class DatabaseLog extends AbstractLog implements Savable {

    private final LogBeanDao mLogBeanDao;
    private DatabaseThresholdWatchDog mDatabaseWatchDog;
    private static DatabaseLog INSTANCE = null;

    public static DatabaseLog create(Context context){
        if (INSTANCE == null){
            INSTANCE = new DatabaseLog(context);
        }
        return INSTANCE;
    }

    private DatabaseLog(Context context) {

        mLogBeanDao = LogBeanDaoImpl.getInstance(context);

        // 开启数据库监测线程
        mDatabaseWatchDog = new DatabaseThresholdWatchDog(mLogBeanDao);
        mDatabaseWatchDog.start();

        // 保存方法实现
        LogCache.get().setSavable(this);
    }

    @Override
    public void destroy() {
        INSTANCE = null;
        mDatabaseWatchDog.interrupt();
        mDatabaseWatchDog = null;
    }

    @Override
    public void save(final List<LogBean> logBeans) {
        mLogBeanDao.save(logBeans);
    }

    /**
     * 周期性检查数据库数据大小是否达到上限
     */
    static class DatabaseThresholdWatchDog extends Thread {

        private final LogBeanDao mLogBeanDao;

        DatabaseThresholdWatchDog(LogBeanDao logBeanDao) {
            super("Check database thread");
            setDaemon(true);
            mLogBeanDao = logBeanDao;
        }

        @Override
        public void run() {
            Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND);
            while (true) {
                try {
                    Thread.sleep(SLog.sDbCheckInterval);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                    break;
                }
                mLogBeanDao.checkThreshold();
            }
        }
    }
}

File: loglib/src/main/java/com/robog/loglib/DatabaseLog.kt
package com.robog.loglib

import android.content.Context
import android.os.Process
import com.robog.loglib.database.LogBeanDao
import com.robog.loglib.database.LogBeanDaoImpl

/**
 * Created by yuxingdong on 2018/6/29.
 * <p>
 * 将Log对象存入数据库，并开启数据库监测，定期检查数据库数据数量是否达到上限。
 * </p>
 */
internal class DatabaseLog private constructor(
        context: Context
) : AbstractLog(), Savable {

    private val logBeanDao: LogBeanDao
    private var databaseWatchDog: DatabaseThresholdWatchDog? = null

    init {

        logBeanDao = LogBeanDaoImpl.getInstance(context)
        databaseWatchDog = DatabaseThresholdWatchDog(logBeanDao)
        databaseWatchDog!!.start()

        LogCache.get().savable = this
    }

    companion object {
        private var instance: DatabaseLog? = null

        fun create(context: Context): DatabaseLog = instance
                ?: DatabaseLog(context).also { instance = it }
    }

    override fun destroy() {
        instance = null
        databaseWatchDog?.interrupt()
        databaseWatchDog = null
    }

    override fun save(logBeans: List<LogBean>) {
        logBeanDao.save(logBeans)
    }

    internal class DatabaseThresholdWatchDog(
            private val innerLogBeanDao: LogBeanDao
    ) : Thread("Check database thread") {

        init {
            isDaemon = true
        }

        override fun run() {
            Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND)
            while (true) {
                try {
                    Thread.sleep(SLog.dbCheckInterval)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                    break
                }
                innerLogBeanDao.checkThreshold()
            }
        }
    }
}